packages/scene-composer/src/components/panels/SceneHierarchyPanel/SceneHierarchyDataProvider.tsx

```diff
@@ -213,7 +213,21 @@ const SceneHierarchyDataProvider: FC<SceneHierarchyDataProviderProps> = ({ selec
         return;
       }
 
-      const newParentObject = getObject3DBySceneNodeRef(newParentRef);
+      let newParentObject = getObject3DBySceneNodeRef(newParentRef);
+      const newParent = getSceneNodeByRef(newParentRef);
+
+      // If the direct new parent is a sub model node, use its modelRef parent
+      // to calculate the new transforms to prevent the object froming being
+      // moved to random position.
+      if (findComponentByType(newParent, KnownComponentType.SubModelRef)) {
+        let hierarchyParent = newParent;
+        while (hierarchyParent) {
+          if (findComponentByType(hierarchyParent, KnownComponentType.ModelRef)) break;
```

Comment:
This if-statement can be removed by updating the while-statement as below (loop until we find a node which is ModelRef):
`while(!findComponentByType(hierarchyParent, KnownComponentType.ModelRef))`

Reply:
Not a bad suggestion, but this should not be a blocking request.

Reply:
+1 - Just that the while-statement conveys its purpose clearly.

Reply:
I will combine the if into while. we still need to check hierarchyParent exists, otherwise, if somehow there isn't a modelref parent until root, we will have infinite loop

Reply:
Sure, we can do that.
Just for my understanding - Is there a case when modelref parent doesn’t exists?